### DISCUSSION TITLE
How would I map Alt+Tab to toggle to the most recent window?

### DISCUSSION THREAD

**Discussion by gmayerw** [xmonad/xmonad#469] [General]
Not trying to emulate Windows here but you have to take advantage of their very few good ideas.

How can I toggle focus between my current window and the last window that was focused? If I press Alt+Tab twice, it should go bring focus back to my current window.

I'd love to see if we can write code to do it per workspace but also globally on all workspaces. I'm not sure which one I want to use yet. Thanks!

**Comment by liskin** [comment]
Personally I use https://github.com/sagb/alttab but https://xmonad.github.io/xmonad-docs/xmonad-contrib-0.17.1.9/XMonad-Actions-MostRecentlyUsed.html is a fairly recent contribution that solves the same problem (haven't tried it myself yet).

**Comment by gmayerw** [comment]
Thank you. I'm on arch linux and alttab looks like an AUR package. I usually try to stay away from those as they don't really seem to be well maintained. 

So what I have is:
```
{-# LANGUAGE NoMonomorphismRestriction #-}
import XMonad hiding ((|||))
import XMonad.Actions.CycleWS
import XMonad.Actions.GridSelect
import XMonad.Actions.MostRecentlyUsed
import XMonad.Layout.LayoutCombinators
import XMonad.Util.EZConfig (additionalKeysP)
import XMonad.Util.WorkspaceCompare
import qualified Data.Map as M
import qualified XMonad.StackSet as W

conf = def
        { modMask = mod4Mask
        , terminal = "kitty -e vifm -c empty $HOME ; rm $HOME/.vim/undodir/* ; rm $HOME/.viminfo"
        , borderWidth = 0
        , layoutHook = Full ||| tiled
        }

myNavigation :: TwoD a (Maybe a)
myNavigation = makeXEventhandler $ shadowWithKeymap navKeyMap navDefaultHandler
 where navKeyMap = M.fromList [
          ((0,xK_Escape), cancel)
         ,((0,xK_space) , select)
         ,((0,xK_slash) , substringSearch myNavigation)
         ,((0,xK_Left)  , move (-1,0)  >> myNavigation)
         ,((0,xK_h)     , move (-1,0)  >> myNavigation)
         ,((0,xK_Right) , move (1,0)   >> myNavigation)
         ,((0,xK_l)     , move (1,0)   >> myNavigation)
         ,((0,xK_Down)  , move (0,1)   >> myNavigation)
         ,((0,xK_j)     , move (0,1)   >> myNavigation)
         ,((0,xK_Up)    , move (0,-1)  >> myNavigation)
         ,((0,xK_k)     , move (0,-1)  >> myNavigation)
         ,((0,xK_y)     , move (-1,-1) >> myNavigation)
         ,((0,xK_i)     , move (1,-1)  >> myNavigation)
         ,((0,xK_n)     , move (-1,1)  >> myNavigation)
         ,((0,xK_m)     , move (1,-1)  >> myNavigation)
         ,((0,xK_Return), setPos (0,0) >> myNavigation)
         ]
       navDefaultHandler = const myNavigation

myColorizer =
  colorRangeFromClassName
    (0x00, 0x00, 0x00) -- lowest inactive bg
    (0x00, 0x00, 0x00) -- highest inactive bg
    (0xff, 0xff, 0xff) -- active bg
    (0xff, 0xff, 0xff) -- inactive fg
    (0x00, 0x00, 0x00) -- active fg

gsconfig1 colorizer =
  (buildDefaultGSConfig myColorizer)
    { gs_cellheight = 50
    , gs_cellwidth = 300
    , gs_cellpadding = 10
    , gs_bordercolor = "black"
    , gs_font = "xft:DejaVuSans-12:regular"
    , gs_navigate = myNavigation
    }

main :: IO ()
main = xmonad . configureMRU $ conf
      `additionalKeysP`
        [ ("M-S-<Space>", gets windowset >>= mapM_ (windows . W.shiftWin "9") . W.allWindows >> windows (W.view "9") >> asks (layoutHook . config) >>= setLayout >> sendMessage (JumpToLayout "Tall"))
        , ("C-S-<Space>", asks (layoutHook . config) >>= setLayout)
        , ("M-;", goToSelected $ gsconfig1 myColorizer)
        , ("M-C-<Space>", withFocused (\w -> windows (W.view "9" . W.shiftWin "9" w) >> doTo Next emptyWS getSortByIndex (\ws -> windows (W.view ws . W.shiftWin ws w))) >> asks (layoutHook . config) >>= setLayout)
        , ("M1-<Tab>", mostRecentlyUsed [xK_Alt_L, xK_Alt_R] xK_Tab) ]

tiled = Tall 1 (3/100) (1/2)
```
And on compilation and restart I get this error:
```
XMonad recompiling (forced).
Errors detected while compiling xmonad config: /home/wgm/.xmonad/xmonad.hs
$ ghc --make xmonad.hs -i -ilib -fforce-recomp -main-is main -dynamic -v0 -outputdir /home/wgm/.xmonad/build-x86_64-linux -o /home/wgm/.xmonad/xmonad-x86_64-linux

xmonad.hs:5:1: error:
    Could not find module ‘XMonad.Actions.MostRecentlyUsed’
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
5 | import XMonad.Actions.MostRecentlyUsed
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Please check the file for errors.

xmonad: xmessage: executeFile: does not exist (No such file or directory)
```

**Comment by geekosaur** [comment]
You need to be running xmonad from git, as we haven't made a release including it yet; see the instructions in https://xmonad.org/INSTALL.html.

**Comment by gmayerw** [comment]
Ok thanks. Do you know about how long until it will be available through the official arch repositories?

**Reply by geekosaur** [reply]
That's up to the Arch maintainer(s), but my guess is not until we make a release and we have no plans at present. That said, xmonad-git is in AUR and pulls directly from git.

**Reply by slotThe** [reply]
> we have no plans at present. That said, xmonad-git is in AUR and pulls directly from git.

Actually, I would like to cut a release before the end of this year (more on that soon, hopefully), but knowing Arch it may well take quite a while before that one lands in their repos

**Reply by gmayerw** [reply]
Awesome! Checking back to see if this release got cut yet. I updated my system and still getting the library not found error. Thanks! @slotThe

**Reply by slotThe** [reply]
Alas, not yet! But we're reasonably close, the only thing that's missing (at least imo) for a new release is to incorporate https://github.com/xmonad/xmonad-contrib/pull/825, which will hopefully be done soon™

**Comment by gmayerw** [comment]
Definitely not a solution, but the closest thing I could find to what I want is:

```
import XMonad.Actions.CycleWS
...
main = xmonad $ conf
      `additionalKeysP`
        [ ...
        , ("M1-<Tab>", toggleWS) ]

```

**Reply by gmayerw** [reply]
This will work if you move the window that you want to toggle to a new workspace.

**Comment by gmayerw** [comment]
Looking like this update might come out this year? I saw an update on arch
linux for xmonad, it may be in there. I'll test it when I get time

**Reply by slotThe** [reply]
@wgmayer0 0.18.0 did indeed land last month! Seems like Arch already packaged that version, so you should be good to go!

**Comment by gmayerw** [comment]
Wow guys... great job on this feature. This is great!! Works perfectly. Thank you!
